fix(zk): properly mark unlinked contract

Karrq opened the pull request:
With this PR we can now properly detect when a contract is unlinked, mark it as such and track the missing libraries associated with the unlinked bytecode.

Changes:
* switch from `evm` to `eravm` object in json output
* remove `methodIdentifiers`
* deserialize `missingLibraries` and store it next to the unlinked bytecode

Karrq committed:
feat(zk): mark bytecode as unlinked

refactor(zk): get output from `eravm` object

Karrq committed:
fix(zk): retain previous artifact layout

elfedy commented on crates/artifacts/zksolc/src/bytecode.rs:
```diff
@@ -1,23 +1,54 @@
+
+    /// This value is not part of the compiler output json
+    ///
+    /// It will be populated if the contract has any missing library
```
```suggestion
    // This value is not part of the compiler output json
    // It will be populated if the contract has any missing library
```

elfedy replied:
Not a fan of mixing deserializable and non deserializable fields in the same struct, but might be personal preference. Maybe we could have different structs for the contract and the artifact. Contract has `Bytecode` with the deserialized fields and artifact has `ArtifactBytecode` with the aggregated data where we have this metadata, then it's artifact bytecode that we implement the conversions to the `Compact` stuff for.

elfedy commented on crates/artifacts/zksolc/src/bytecode.rs:
```diff
@@ -1,23 +1,54 @@
     fn from(bcode: Bytecode) -> Self {
-        Self { object: bcode.object, source_map: None, link_references: BTreeMap::default() }
+        let link_references = bcode.link_references();
+        Self { object: bcode.object, source_map: None, link_references }
```
We need to be careful now because bytecode and deployed bytecode will NOT be the same thing as  bytecode might be an unlinked and `link_references` has meaningful info . Maybe update the `NOTE` above.
Also is having this conversion the reason why you need `missing_libraries` in the `Bytecode` struct

Karrq replied:
in EraVM they still are the same thing, not sure why the unlinked means it's different

elfedy replied:
`CompactBytecode` will have the link references for the bytecode `CompactDeployedBytecode` will not. Previously we only cared about the bytecode field and could use both interchangebly. It may not be the case now if we used the `linked_references` somewhere we need `CompactBytecode`

elfedy commented on crates/artifacts/zksolc/src/contract.rs:
```diff
@@ -8,10 +8,30 @@ use foundry_compilers_artifacts_solc::{
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[serde(transparent)]
+pub struct Contract(
```
Why do you need this wrapper?

Karrq replied:
to put `missing_libraries` in the right place

elfedy commented on crates/artifacts/zksolc/src/lib.rs:
```diff
@@ -91,22 +92,18 @@ impl CompilerOutput {
     pub bytecode: Option<Bytecode>,
-    /// The list of function hashes
-    #[serde(default, skip_serializing_if = "::std::collections::BTreeMap::is_empty")]
-    pub method_identifiers: BTreeMap<String, String>,
```
Are these not output anymore?

Karrq replied:
not in `eravm`

elfedy commented on crates/artifacts/zksolc/src/serde_helpers.rs:
```diff
@@ -0,0 +1,61 @@
+
+/// Deserialize a [`RawContract`] with added `missingLibraries` field,
+/// to populate [`Bytecode`]'s `missing_libraries` with it
+pub fn maybe_unlinked_contract<'de, D>(deserializer: D) -> Result<RawContract, D::Error>
```
Maybe do this when converting from contract to artifact and not when deserializing

Karrq replied:
That sounds good

elfedy commented on crates/artifacts/zksolc/src/serde_helpers.rs:
```diff
@@ -0,0 +1,61 @@
+
+/// Deserialize a [ `Bytecode` ] by either the raw bytecode object (from the compiler output),
+/// or from a wrapped bytecode object from a stored artifact
+pub fn opt_maybe_unwrapped_bytecode<'de, D>(deserializer: D) -> Result<Option<Bytecode>, D::Error>
```
Same I would prefer having separate structs for both and not join them in a single one

elfedy reviewed (commented):
Unless I am missing something on why it is done this way, I think we should consider separating compiler outputs from the artifacts we use and not join them into one, hiding the conversion inside the deserialization. The abstractions for this are there already:
1) We deserialize `Contract` that has a `Bytecode` and we respect the output fields (no ad hoc deserialization logic).
2) We have `ZkContractArtifact` with  `ZkArtifactBytecode` that contains the `missing_libraries` (if we need them there for some reason) and that is constucted on the  `zksync_contract_to_artifact_function`

Karrq committed:
feat(artifacts:zk): bytecode artifact format

fix: remove serde logic in favor of keeping artifact and compiler format separate

hedgar2017 commented on crates/artifacts/zksolc/src/contract.rs:
```diff
@@ -31,12 +31,70 @@ pub struct Contract {
+        self.hash.is_none() || !self.missing_libraries.is_empty()
+    }
+
+    pub fn parse_link_references(
```
Rename this function?

elfedy commented on crates/compilers/src/compilers/zksolc/mod.rs:
```diff
@@ -728,7 +729,7 @@ mod tests {
         let (_, mut contracts) = out.split();
         let contract = contracts.remove("LinkTest").unwrap();
-        let bytecode = &contract.evm.unwrap().bytecode.unwrap().object;
+        let bytecode = &contract.0.eravm.unwrap().bytecode.unwrap().object;
```
```suggestion
        let bytecode = &contract.eravm.unwrap().bytecode.unwrap().object;
```
There's no wrapper anymore right?

elfedy commented on crates/compilers/src/compilers/zksolc/mod.rs:
```diff
@@ -741,7 +742,7 @@ mod tests {
         let (_, mut contracts) = out.split();
         let contract = contracts.remove("LinkTest").unwrap();
-        let bytecode = &contract.evm.unwrap().bytecode.unwrap().object;
+        let bytecode = &contract.0.eravm.unwrap().bytecode.unwrap().object;
```
```suggestion
        let bytecode = &contract.eravm.unwrap().bytecode.unwrap().object;
```

Karrq replied:
weird, I thought I committed the fix already

elfedy commented on crates/compilers/src/zksync/artifact_output/zk.rs:
```diff
@@ -24,18 +24,19 @@ use std::{
-    //#[serde(default, skip_serializing_if = "Vec::is_empty")]
-    //pub generated_sources: Vec<GeneratedSource>,
+    // #[serde(default, skip_serializing_if = "Option::is_none")]
+    // pub method_identifiers: Option<BTreeMap<String, String>>,
```
If we don't use this maybe delete

elfedy commented on crates/compilers/src/zksync/artifact_output/zk/bytecode.rs:
```diff
@@ -0,0 +1,35 @@
+    }
+}
+
+// NOTE: distinction between bytecode and deployed bytecode make no sense of zkEvm, but
```
```suggestion
// NOTE: distinction between bytecode and deployed bytecode makes no sense of zkEvm, but
```

elfedy reviewed (commented):
LGTM, ci is not passing though, I think it is due to outdated compiler tests.

Karrq committed:
chore: cleanup
